app/gui/language/span-tree/src/generate.rs

```diff
@@ -448,63 +528,230 @@ fn generate_node_for_prefix_chain<T: Payload>(
+    let mut application = app_base.resolve(context);
 
     // When using method notation, expand the infix access chain manually to maintain correct method
     // application info and avoid generating expected arguments twice. We cannot use the
     // `generate_node` implementation on `GeneralizedInfix`, because it always treats the
     // access chain as a method call without any arguments applied.
     let node = if let Some(infix) = GeneralizedInfix::try_new(&this.func) {
-        generate_node_for_opr_chain(infix.flatten(), node::Kind::Operation, app_base, context)
+        generate_node_for_opr_chain(
+            infix.flatten(),
+            node::Kind::Operation,
+            app_base,
+            application.as_mut(),
+            context,
+        )
     } else {
         this.func.generate_node(node::Kind::Operation, context)
     };
 
-    let ret = this.args.iter().enumerate().fold(node, |node, (i, arg)| {
-        let node = node?;
-        let is_first = i == 0;
-        let is_last = i + 1 == prefix_arity;
-        let arg_kind = if is_first && !uses_method_notation {
-            node::Kind::from(node::Kind::this().with_removable(removable))
-        } else {
-            node::Kind::from(node::Kind::argument().with_removable(removable))
-        };
+    let argument_positions = resolve_argument_positions(&this.args, application);
+    let num_positions = argument_positions.len();
+
+    let mut inserted_arguments = 0;
+
+    argument_positions
+        .into_iter()
+        .enumerate()
+        .fold(node, |node, (i, position)| {
+            let node = node?;
+            let is_first = i == 0;
+            let is_last = i == num_positions - 1;
+            let is_only_argument = is_first && is_last;
+
+            match position {
+                ArgumentPosition::ChainArgument { arg, named, info } => {
+                    let mut gen = ChildGenerator::default();
+                    gen.add_node(vec![PrefixCrumb::Func.into()], node);
+                    gen.spacing(arg.sast.off);
+
+                    let arg_name = named.as_ref().map(|named| named.name);
+
+                    // Allow removing the prefix chain arguments only if it doesn't destroy the
+                    // chain or influence the argument logical positions. We can always remove
+                    // resolved arguments, as the remaining positions will be rewritten using known
+                    // argument names and placeholders will be inserted. For unresolved arguments,
+                    // we can remove them only if removal preserves the prefix application.
+                    let resolved = info.is_some();
+                    let removable = resolved || !is_only_argument;
+
+                    let mut arg_kind = node::Kind::from(
+                        node::Kind::argument()
+                            .with_removable(removable)
+                            .with_name(arg_name.map(ToString::to_string)),
+                    );
+
```

Comment:
this is pretty long fn, are we able to split it into several Fns? No worries if not

Reply:
I tried, but it only led to scattering of code that has a single place and function. I think it is easier to follow as a single contiugous block.

Reply:
ok!